### PR TITLE
Change the implementation of the priority queue to use a heap data structure

### LINKED ISSUE #103: "EventQueue size exceeded" exception
I'm trying to build a large RAM component from smaller building blocks and eventually the simulator hits an "EventQueue size exceeded" exception.

Once this happens the editor goes into a bad state, failing to draw new components and (perhaps) even save the existing work.

I suppose this limit might exist to keep the simulation contained; if that is the case, perhaps the error should be handled in a graceful way so the circuit doesn't get wasted.

Alternatively, we could increase the limit of this queue or make it configurable on a per project basis.

For a repro, see this circuit: https://circuitverse.org/users/1935/projects/4774
In the 16KB circuit (which is incomplete) I'm trying to add another 4KB circuit; doing so causes the exception.

### PR DESCRIPTION
**Pull request by Prerna-0202**
### **GSOC'23 TASK**

Fixes #103 

#### Describe the changes you have made in this PR -

 1. I have changed the implementation of the priority queue to use a heap data structure. Previously, it was using a linear data structure. 
Now, It is making use of a binary heap data structure to manage the priority queue. The add() method uses heapify-up to find the correct position for adding an element, and the remove() method uses heapify-down to maintain the heap property after removing an element. These heapify operations have logarithmic time complexity (O(log n)), which makes the binary heap implementation much more efficient for managing a priority queue compared to the linear implementation.
2. Now, as the time complexity reduces to (O(log n)) from (O(n)), I have changed the simulation queue limit to 100000.

Screenshots:-

Before:-

Circuit was not  able to load properly and freezes the simulator

Screenshot (1774)

https://github.com/CircuitVerse/CircuitVerse/assets/89515816/91ff4eac-59ff-44d0-9f52-b777f5f66866

After:-

Screenshot (1772)

Screenshot (1771)

https://github.com/CircuitVerse/CircuitVerse/assets/89515816/043172e3-640d-4049-9c9f-8a2759909fe1

Note: Please check **Allow edits from maintainers.** If you would like us to assist in the PR.

### CODE REVIEW

**Review by gr455 — CHANGES_REQUESTED** [review]
@Prerna-0202 nit: Can you change your naming of `this.frontIndex` to `this.backIndex`.
The front of the queue is where you pop elements from. In your case it seems you are popping from 0, so 0 is the frontIndex. You add elements through the back (although it doesnt matter in a PQ since even in the earlier implementation, the objs were being added from the front).

But yeah, naming it backIndex will be clearer in my opinion.

**Comment by gr455** [comment]
Everything else looks good! Great job with this

**Comment by gr455** [comment]
// @satu0king Please review

File: simulator/src/eventQueue.js [hunk 1]
```diff
@@ -73,31 +66,70 @@ export default class EventQueue {
...
+            index = parentIndex;
+        }
+    }
+
+    /**
+    * Function to heapify the element down to its correct position.
+    * @param {number} index
+    */
+    heapifyDown(index) {
+        while (true) {
```
[the remark this answers is not in the document]
Comment (gr455) [reply simulator/src/eventQueue.js:88]: Ignoring

File: simulator/src/eventQueue.js [hunk 2]
```diff
@@ -73,31 +66,70 @@ export default class EventQueue {
...
+            const leftChildIndex = 2 * index + 1;
+            const rightChildIndex = 2 * index + 2;
+            let smallestChildIndex = index;
+
+            if (leftChildIndex < this.frontIndex && this.queue[leftChildIndex].queueProperties.time < this.queue[smallestChildIndex].queueProperties.time) {
```
[the remark this answers is not in the document]
Comment (gr455) [reply simulator/src/eventQueue.js:93]: Ignoring

File: simulator/src/eventQueue.js [hunk 3]
```diff
@@ -73,31 +66,70 @@ export default class EventQueue {
...
+                smallestChildIndex = leftChildIndex;
+            }
+
+            if (rightChildIndex < this.frontIndex && this.queue[rightChildIndex].queueProperties.time < this.queue[smallestChildIndex].queueProperties.time) {
```
[the remark this answers is not in the document]
Comment (gr455) [reply simulator/src/eventQueue.js:97]: Ignoring

**Comment by Prerna-0202** [comment]
> @Prerna-0202 nit: Can you change your naming of `this.frontIndex` to `this.backIndex`. The front of the queue is where you pop elements from. In your case it seems you are popping from 0, so 0 is the frontIndex. You add elements through the back (although it doesnt matter in a PQ since even in the earlier implementation, the objs were being added from the front).
> 
> But yeah, naming it backIndex will be clearer in my opinion.

yess, Done 👍

**Review by satu0king — APPROVED** [review]

**Review by gr455 — APPROVED** [review]

**Pull request merged**